doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+    struct HasAllocateMethod : std::false_type {};
+
+    template <typename C>
+    struct HasAllocateMethod<
+        C, std::void_t<std::enable_if_t<
+               std::is_same<decltype(std::declval<C>().allocate(
+                                std::declval<uint64_t>())),
+                            void>::value,
+               void>>> : std::true_type {};
+
+    // test all other methods
+    static_assert(HasAllocateMethod<T>::value, 
+              "Type is not an allocator, void* allocate(uint64_t) missing");
+};
+```
+
+## Design of Zero Copy Containers
+
+The design pursues the following goals.
+
+1. The `cxx::vector<T, Capacity>` should not change in its API.
+
+2. All static cxx containers are still owner of all their data. This means the
+   memory which the stack allocator manages must be stored inside the cxx
+   container class.
+
+3. A function developer should not have to specify the allocator type of a
+   container. This means the allocator type is independent of the container type.
+   This has to work:
+   ```cpp
+   void myFunkyFunction(const cxx::vector<int> & bla) {}
+
+   cxx::vector<int, 20> a;
+   cxx::vector<int, 30> b;
+   cxx::vector<int> c(Allocator::heap);
+
+   myFunkyFunction(a);
+   myFunkyFunction(b);
+   myFunkyFunction(c);
+   ```
+
+4. All containers must be shared memory compatible, this means no vtables and
+   virtual. Furthermore the usage of function pointers, `cxx::function` and
+   `cxx::function_ref` is forbidden in all containers and their underlying 
+   constructs.
+
+5. Before using any cxx container the user as to call `reserve()` once otherwise
+   the container will have a default capacity of zero. It is not allowed to
+   call `reserve()` multiple times since this could violate the zero copy
+   guarantee.
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+   * It is not possible to provide solely the capacity (`reserve` parameter) via
+     the constructor of the container since frameworks like ROS2 cannot forward
+     such arguments when for instance `loan` was called. But it is possible to
+     provide it as an alternative.
+
+6. The allocator type should not change the underlying container type, e.g. the
+   allocator should not be provided as template argument, otherwise
+   it becomes impossible to reuse defined structures in a non shared memory
+   context. This is for instance required when a user would like to store some
+   received data locally in a cache for later use.
+
+7. The allocator should not be part of copy and move-assignment operations.
+   The container should behave as closely as possible to an STL container with
+   custom allocator.
+   ```cpp
+   cxx::vector a(HeapAllocator);
+   cxx::vector b(SharedMemoryAllocator);
+   cxx::vector c(AnotherAllocator);
+   cxx::vector d(HeapAllocator);
+
+   a = b; // the content is copied from shared memory into the heap
```

Comment:
I think the behavior needs to be clarified more here. I think `a` will/should deallocate its current memory (or extend it), allocate more memory if needed and then copy the data from b to this memory.

However, in general this operation needs to know something about the source memory here (and it will, since it has access to the allocator of the source `b`).

TLDR: the actual semantics need more detail here I think.